src/Stratis.Bitcoin.Features.PoA/Voting/VotingDataEncoder.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Features.PoA.Voting
+{
+    public class VotingDataEncoder
+    {
+        /// <summary>Prefix used to identify OP_RETURN output with voting data.</summary>
+        public static readonly byte[] VotingOutputPrefixBytes = new byte[] { 143, 18, 13, 254 };
+
+        public const int VotingDataMaxSerializedSize = ushort.MaxValue;
+
+        private readonly ILogger logger;
+
+        public VotingDataEncoder(ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>Decodes raw voting data.</summary>
+        /// <exception cref="PoAConsensusErrors.VotingDataInvalidFormat">Thrown in case voting data format is invalid.</exception>
+        public List<VotingData> Decode(byte[] votingDataBytes)
+        {
+            try
+            {
+                if (votingDataBytes.Length > VotingDataMaxSerializedSize)
+                {
+                    this.logger.LogTrace("(-)[INVALID_SIZE]");
+                    PoAConsensusErrors.VotingDataInvalidFormat.Throw();
+                }
+
+                var memoryStream = new MemoryStream(votingDataBytes);
+                var deserializeStream = new BitcoinStream(memoryStream, false);
+
+                var decoded = new List<VotingData>();
+
+                deserializeStream.ReadWrite(ref decoded);
+
+                return decoded;
+            }
+            catch (Exception e)
```

Comment:
Its normally not common to catch all exceptions in consensus rules, need to discuss.

Reply:
All exceptions here will be serialization exceptions. 
And it's hard to say exactly what those would be.

So in order to avoid consensus stopping in case we get it I went with suppressing all exceptions and re-throwing them as a consensus exception.

Reply:
Do we have any other examples of deserializing a type in the rule validation?
I guess this is fine if you rethrow (it will be determinitic anyway)